airflow/executors/inprocess_executor.py

```diff
@@ -0,0 +1,152 @@
+    """
+
+    _terminated = threading.Event()
+
+    def __init__(self):
+        super().__init__()
+        self.tasks_to_run: List[TaskInstance] = []
+        # Place where we keep information for task instance raw run
+        self.tasks_params: Dict[TaskInstanceKeyType, Dict[str, Any]] = {}
+        self.fail_fast = conf.getboolean("debug", "fail_fast")
+
+    def execute_async(self, *args, **kwargs) -> None:
+        """
+        The method is replaced by custom trigger_task implementation.
+        """
+
+    def sync(self) -> None:
+        task_failed = False
+        while self.tasks_to_run:
+            ti = self.tasks_to_run.pop(0)
+            if self.fail_fast and task_failed:
+                self.log.info("Setting %s to %s", ti.key, State.UPSTREAM_FAILED)
+                ti.set_state(State.UPSTREAM_FAILED)
+                self.change_state(ti.key, State.UPSTREAM_FAILED)
+                continue
+
+            if self._terminated.is_set():
+                self.log.info(
+                    "Executor is terminated! Stopping %s to %s", ti.key, State.FAILED
+                )
+                ti.set_state(State.FAILED)
+                self.change_state(ti.key, State.FAILED)
+                continue
+
+            task_failed = self._run_task(ti)
+
+    def _run_task(self, ti: TaskInstance) -> bool:
+        self.log.debug("Executing task: %s", ti)
+        key = ti.key
+        try:
+            params = self.tasks_params.get(ti.key, {})
+            ti._run_raw_task(  # pylint: disable=protected-access
+                job_id=ti.job_id, **params
+            )
+            self.change_state(key, State.SUCCESS)
+            return False
+        except Exception as e:  # pylint: disable=broad-except
+            self.change_state(key, State.FAILED)
+            self.log.exception("Failed to execute task: %s.", str(e))
+            return True
+
+    def queue_task_instance(
+        self,
+        task_instance: TaskInstance,
+        mark_success: bool = False,
+        pickle_id: Optional[str] = None,
+        ignore_all_deps: bool = False,
+        ignore_depends_on_past: bool = False,
+        ignore_task_deps: bool = False,
+        ignore_ti_state: bool = False,
+        pool: Optional[str] = None,
+        cfg_path: Optional[str] = None,
+    ) -> None:
+        """
+        Queues task instance with empty command because we do not need it.
+        """
+        self.queue_command(
+            task_instance,
+            [str(task_instance)],  # Just for better logging, it's not used anywhere
+            priority=task_instance.task.priority_weight_total,
+            queue=task_instance.task.queue,
+        )
+        # Save params for TaskInstance._run_raw_task
+        self.tasks_params[task_instance.key] = {
+            "mark_success": mark_success,
+            "pool": pool,
+        }
+
+    def trigger_tasks(self, open_slots: int) -> None:
```

Comment:
we are copy pasting some code here with some business logic from baseexecutor, 
 I am not sure if there is another workaround here, we need TI to manage it state

let's see how it goes through code review

Reply:
Yes it's mostly workaround for type annotations. Previously when executors where untyped it didn't matter what we queued.